### DISCUSSION TITLE
Wiring limit switches

### DISCUSSION THREAD

**Discussion by jds-designs** [gizmo-platform/gizmo#24] [Q&A] [answered]
Hello, I need some clarification on how to wire limit switches for the Gizmo board. The old VEX Cortex system allowed limit switches to be connected with two wires, as shown below:
![VEX Cortex limit switch small](https://github.com/user-attachments/assets/2196b065-41fd-4305-a5e3-1d846d8c286f)

But I saw in a video on BESTedu that the instructor wired a limit switch with three wires:
![Screenshot 2024-09-12 203229](https://github.com/user-attachments/assets/3fc60de4-89df-4e15-9a2e-a9723a4ca6e0)
I also read the [BESTedu Getting Started lesson on switches](https://bestedu.bestrobotics.org/courses/best-gizmo-getting-started/lessons/using-microswitches/), but I wasn't clear whether or not it was saying to connect all three wires to all three lugs on the switch.

I apologize if this question has an obvious answer I missed; I just want to make sure I'm doing everything correctly as I learn this new system.

Thank you!

**Comment by barulicm** [comment, marked as answer]
That is a very reasonable question. You can use either wiring technique.

If you use the two-wire technique, the switch will leave the port in a disconnected, or "floating", state. In this state nothing is driving the port either high or low, so the signal will effectively be random. You'll need to use the Gizmo's internal pull-up resistors, which are activated through software. The exact method will depend on which programming language you are using.

In Arduino, when calling the `pinMode` function, set the mode to `INPUT_PULLUP`.

```C++
pinMode(switchPin, INPUT_PULLUP);
```

In CircuitPython, you can turn on the pull up resistor by setting the `pull` property on the digitalio object. Example:
```Python
switch = digitalio.DigitalInOut(board.SLIDE_SWITCH)

# using switch_to_input function
switch.switch_to_input(pull=digitalio.Pull.UP)

# Or, using properties
switch.direction = digitalio.Directoin.INPUT
switch.pull = digitalio.Pull.UP
```

Using the three-wire configuration avoids the need for pull-up resistors because the switch itself drives the port in both the pressed and released states.

**Reply by the-maldridge** [reply]
Its worth noting that while you can use the 3-wire method safely with the Gizmo, you should always consult the MCU documentation for the chip you're using.  Connecting the controller in this way technically has no current limiting which can damage some microcontrollers.  The Raspberry Pi Pico can tolerate this, but not all controllers can.

**Reply by jds-designs** [reply]
Thank you both for the information.

I realized that the old 2-wire interface cables can only connect to the GPIO 6 and the GND rows, skipping the VDC (3.3v) row. Would a limit switch function if wired this way?
![2-wire cable](https://github.com/user-attachments/assets/e2344d21-9f19-4821-ad23-bccc225e1adc)

**Reply by barulicm** [reply]
Yes. When the switch is open, the pull-up resistor will hold the port high (reporting true in software). When the switch closes, it will pull the port to ground (reporting false in software).

**Reply by jds-designs** [reply]
Ok, thanks. Just to make sure I understand how it works - the GPIO and GND pin sockets are what open and close the circuit, returning a signal of HIGH or LOW (1 or 0 in the code). The VDC (3.3v) socket drives current to the port so it will not be disconnected, preventing it from fluctuating in voltage & giving random signals. Is that correct?

**Comment by jds-designs** [comment]
Thanks for the clarification. Most of the returnable kits at my hub have the two-wire interface cables, so I will definitely want to pass this info on to the teachers and students at our kickoff next weekend.

To that end, I've created some diagrams of the different ways to wire limit switches to the Gizmo. Please let me know if you happen to see any errors in them.
![Switch Normally Open 3-wire](https://github.com/user-attachments/assets/42069277-c69b-4bcb-bb24-c91fc2edfe62)
![Switch Normally Closed 3-wire](https://github.com/user-attachments/assets/95268759-c200-4d08-90ce-12204394f97f)
![Switch Normally Open 2-wire](https://github.com/user-attachments/assets/4f15b950-bec5-4e8c-9292-23b9d09c6224)
![Switch Normally Closed 2-wire](https://github.com/user-attachments/assets/4d4ffb81-ea37-4f19-9939-f39442fc46c1)

Thank you!

**Reply by jds-designs** [reply]
Assuming there are no errors in my wiring guide photos, I put them together in a PDF document for easier viewing. I'm posting it here in case anyone else wants to reference it.
[Limit Switch Wiring Diagrams.pdf](https://github.com/user-attachments/files/17004327/Limit.Switch.Wiring.Diagrams.pdf)

**Reply by Vaelatern** [reply]
I see no issues in your diagrams.

I would also encourage you to put, with your sample code, a simple way to tell what the state of each pin is, to reduce any confusion around Normally Open vs. Normally Closed

**Reply by barulicm** [reply]
It's probably worth tweaking the circuitpython code so it attaches the digitalio object to the port being used in the diagrams:

```Python
switch = digitalio.DigitalInOut(gizmo.GPIO_1)
```

(The code I had given earlier was copied from the generic CircuitPython docs, so it doesn't use an actual Gizmo port.)

**Reply by jds-designs** [reply]
Thank you for the feedback. I've made an updated version of the diagram document:
[Limit_Switch_Wiring_Diagrams_v2.pdf](https://github.com/user-attachments/files/17014212/Limit_Switch_Wiring_Diagrams_v2.pdf)
